spec/outputs/s3_spec.rb

```diff
@@ -1,6 +1,318 @@
+      s3.register
+    end
+
+    it "should always delete the source file" do
+      tmp = Stud::Temporary.file
+
+      allow(File).to receive(:zero?).and_return(true)
+      expect(File).to receive(:delete).with(tmp)
+
+      s3.move_file_to_bucket(tmp)
+    end
+
+    it 'should not upload the file if the size of the file is zero' do
+      temp_file = Stud::Temporary.file
+      allow(temp_file).to receive(:zero?).and_return(true)
+
+      expect(s3).not_to receive(:write_on_bucket)
+      s3.move_file_to_bucket(temp_file)
+    end
+
+    it "should upload the file if the size > 0" do
+      tmp = Stud::Temporary.file
+
+      allow(File).to receive(:zero?).and_return(false)
+      expect(s3).to receive(:write_on_bucket)
+
+      s3.move_file_to_bucket(tmp)
+    end
+  end
+
+  describe "#restore_from_crashes" do
+    it "read the temp directory and upload the matching file to s3" do
+      s3 = LogStash::Outputs::S3.new(minimal_settings.merge({ "temporary_directory" => "/tmp/logstash/" }))
+
+      expect(Dir).to receive(:[]).with("/tmp/logstash/*.txt").and_return(["/tmp/logstash/01.txt"])
+      expect(s3).to receive(:move_file_to_bucket_async).with("/tmp/logstash/01.txt")
+
+
+      s3.restore_from_crashes
+    end
+  end
+
+  describe "#receive" do
+    it "should send the event through the codecs" do
+      data = {"foo" => "bar", "baz" => {"bah" => ["a","b","c"]}, "@timestamp" => "2014-05-30T02:52:17.929Z"}
+      event = LogStash::Event.new(data)
+
+      expect_any_instance_of(LogStash::Codecs::Line).to receive(:encode).with(event)
+
+      s3 = LogStash::Outputs::S3.new(minimal_settings)
+      allow(s3).to receive(:test_s3_write)
+      s3.register
+
+      s3.receive(event)
+    end
+  end
+
+  describe "when rotating the temporary file" do
+    before { allow(File).to receive(:delete) }
+
+    it "doesn't skip events if using the size_file option" do
+      Stud::Temporary.directory do |temporary_directory|
+        size_file = rand(200..20000)
+        event_count = rand(3000..15000)
+
+        config = %Q[
+        input {
+          generator {
+            count => #{event_count}
+          }
+        }
+        output {
+          s3 {
+            size_file => #{size_file}
+            codec => line
+            temporary_directory => '#{temporary_directory}'
+            bucket => 'testing'
+          }
+        }
+        ]
+
+        pipeline = LogStash::Pipeline.new(config)
+
+        pipeline_thread = Thread.new { pipeline.run }
+        sleep 0.1 while !pipeline.ready?
+        pipeline_thread.join
+
+        events_written_count = events_in_files(Dir[File.join(temporary_directory, 'ls.*.txt')])
+        expect(events_written_count).to eq(event_count)
+      end
+    end
+
+    it "doesn't skip events if using the time_file option", :tag => :slow do
+      Stud::Temporary.directory do |temporary_directory|
+        time_file = rand(5..10)
+        number_of_rotation = rand(4..10)
+
+        config = {
```

Comment:
I tried using the pipeline for both, but for `Timed constraints` it wasn't possible or I don't know how and using the pipeline make sense for the first one. I will leave them like this.